### DISCUSSION TITLE
Howto send the state of a switch to mqtt-server with HASPmota

### DISCUSSION THREAD

**Discussion by waltkind** [arendst/Tasmota#20222] [Support request] [answered]
Sorry to bother you again with a beginners question, but the learning curve is steep. After searching for hours now I give up and ask here.
I managed to get my screen nice and show the values from the mqtt-server formatted right. In the other direction I succeeded in sending the events of my switch to the mqtt-server. Via the command `tasmota.add_rule("hasp#p2b10#event", / value -> mqtt.publish('tele/lanbon_1/INFO4', value))
` it shows: 'down', 'changed', 'release' and 'up'.
The display shows that the switch has changed and is in another state.
Changing the command from "...#event" to "...#state" didn't work.
What is the command to get the on - off state to the mqtt-server?

**Comment by sfromis** [comment]
If you mean changing the code in your publish to transform values from down/up to off/on, you could do something like:
`mqtt.publish('tele/lanbon_1/INFO4', value=='down' ? 'off' : value=='up' ? 'on' : ''))`
This will publish an empty string if it is neither on nor off. You could also wrap the whole publish in a function to only do a publish if it became either on or off.

Also, I'd recommend using something different from `INFO4` in the topic, as this may be "confusingly similar" to already used topics in Tasmota.

**Comment by waltkind** [comment]
I am aware of mapping, but that is not the problem. 
The events 'down', 'changed', 'release' and 'up' are momentary during the touch of the screen and always end with 'up' when I'm finished. 
But then the state of the switch has changed to either 'on' or 'off' and is shown in color 'bg_color' or 'bg_color10'. This state I want to publish to mqtt.
(The topic `INFO4` is only for test purpose. Will be changed to `Switch1` and so on.)

**Comment by waltkind** [comment]
It is so obvious that I didn't find it. I don't have to test for `state` but for `val`.

But now I have another problem. The line 
`tasmota.add_rule("hasp#p2b15#val", / value -> print("Button p2b15 pressed", value))` 
results in
 `15:22:43.373 Button p2b15 pressed true`.
and after switching again it results in 'false'.

But the line 
`tasmota.add_rule("hasp#p2b15#val", / value -> mqtt.publish('tele/lanbon_1/INFO4', value))` 
results in

> (15:22:43.377 stack traceback:
15:22:43.379 	<native>: in native function
15:22:43.380 	<unknown source>: in function `<lambda>`
15:22:43.381 	<unknown source>: in function `try_rule`
15:22:43.393 	<unknown source>: in function `exec_rules`
15:22:43.395 	<unknown source>: in function `event`
15:22:43.396 	<native>: in native function
15:22:43.407 	<unknown source>: in function `<lambda>`
15:22:43.408 	<unknown source>: in function `run_deferred`
15:22:43.410 	<unknown source>: in function `event`
15:22:43.421 BRY: stack[1] = type='instance' (Tasmota)
15:22:43.422 BRY: stack[2] = type='function' ()
15:22:43.423 BRY: stack[3] = type='instance' (Tasmota)
15:22:43.435 BRY: stack[4] = type='string' (rule)
15:22:43.436 BRY: stack[5] = type='string' ()
15:22:43.437 BRY: stack[6] = type='int' ()
15:22:43.438 BRY: stack[7] = type='string' ({"hasp":{"p2b15":{"event":"changed","val":true}}})
15:22:43.451 BRY: stack[8] = type='string' (type_error)
15:22:43.452 BRY: stack[9] = type='nil' ())`

**Reply by sfromis** [reply, marked as answer]
The payload argument of `mqtt.publish` should be of type string or bytes, and I suspect that you're getting a boolean. If you just want the raw string value, you could use `str(value)`, or a ternary operator similar to what I mentioned previously.

**Reply by waltkind** [reply]
That's it. `str(value)` did the trick. Thanks

**Comment by patpommi** [comment]
Hello, i want to control a lamp with a switch. Publish over zigbee2mqtt. With two buttons (on, off) it works:

"hasp#p1b51#event=up DO publish z2m/L_Schreib/set {"brightness":155}"

But not with a switch. I think thats the same problem.
Could you please post a complete line how to make this?
I have tried various options, but no success.

Thank you so mutch

**Reply by s-hadinger** [reply]
Although I haven't tested, I believe the trigger should be `hasp#p1b51#event=changed` and then you have a boolean (true/false) in `p1b51.val`. I'm not sure however how to craft it in a single rule.

Some simple berry scripting might be better suited so you can adapt to your exact need

**Reply by sfromis** [reply]
It can also be useful with a rule (in Berry or classic) triggering on just `hasp#p1b51#event` allowing you to see what events you get. Or even `hasp#?#?` getting you all the `hasp` events generated.

**Comment by patpommi** [comment]
Many thanks to both of you!
You guys are doing such a great job with this "Tasmota stuff"!!!

The (classic) rules hasp#p1b51#event and hasp#?#? to get the events work. But as output there is only "down, release, up, long and hold".
It is nearly the same as in the post by waltkind.
I was hoping to be able to query the status of the switch. Like on/off or true/false.
But that didn't work.
If there would be a solution for this (with example also for beginners - gladly also in Berry!)
I hope I'm not being too annoying, but this is one of my biggest difficulties in getting a display to work properly.

Wish you a good evening.....

**Reply by sfromis** [reply]
Depends on what you mean by "query". What you get in the events is when something happened, like you pushing, or holding. It is not directly a state, but still you could derive a state.

However, in Berry the variable `p1b51.state` should be 0/1, depending on the position of the toggle. This is closest to the word "query", as it is not about an event happening, but allowing your code can see current state, and do whatever. You could do that when getting triggered by the event of something being done to the switch widget.

Also, did you try what waltkind found, a trigger on `hasp#p2b15#val` ? Looks like getting true/false.

**Comment by patpommi** [comment]
Hello!
Sorry for my late reply, but little time is another problem of mine.
The suggested rule "p1b51.state" or "p1b51#state" gives no output in the console.

Same goes for the suggestions "hasp#p2b15#val" from waltkind and "hasp#p1b51#event=changed".

Thank you for your time. But for now I will use two buttons instead of one switch. I can solve that with what I know so far.
I hope i will dig deeper into the subject and be able to solve it with a switch among other things. Especially with Berry.

**Reply by sfromis** [reply]
No one suggested a **rule trigger** `p1b51.state` or `p1b51#state`. What I suggested was a global variable in Berry named `p1b51.state` as long as that specific object, id 51 on page 1 (in the jsonl file), is what you want. That `p1b51#state` seems to be entirely a figment of your imagination.

And `hasp#p2b15#val` was what waltkind mentioned as an example, cannot make sense unless you point at your specific object 51, obviously not the number which applied to waltkind. I've not tried that, but seems to have worked for waltkind.

Using `hasp#p1b51#event=changed` would only be plausible if you get the value `changed` when triggering on `hasp#p1b51#event` to see what you get.

**Comment by patpommi** [comment]
Sorry, as a beginner I have misunderstood something.
Now I have a few things to try out.
But of course I have adapted the numbers hasp#p2b15# to my situation. ;-)
